### ISSUE TITLE
Teleport movement with laser pointer

### ISSUE THREAD

**Issue by nucleiis** [praydog/UEVR#54]
Thank you for UEVR, it is already great but adding teleport will make it greater because there are tons of people cannot overcome or adapt VR nausea.

Since most UE4/5 games use the same character movement component, I guess it would be quite possible to add some laser pointer which can be controlled via VR controller and event handlers overriding character position / rotation using raycast on the floor (i.e UCharacterMovementComponent::FindFloor).

For VR newbies, this will make UEVR much more comfortable and enjoyable!

**Comment by praydog** [comment]
@narknon thoughts? Do all UE games have consistent raycast flags that are always going to hit a surface correctly?

**Comment by praydog** [comment]
Though if it is part of a FindFloor function for a built-in engine movement component that seems like it would work

**Comment by narknon** [comment]
I think FindFloor is more for directly under the movement component.  Most games do use a child class of the UE movement component, but some don't so that may not always work.  

I think the ideal setup would be either the projectile slope calculation function, two line traces for max tp distance then down to floor, or a single linetrace straight out from pointer if not doing a max distance option (having max distance feels like it would allow for better mods overall so you can't just tp across the map).  

To check if it is a valid TP location you can check if there is nav data on the spot, though that should probably be a toggle for whether to check it or not because some games use weird custom nav.

**Comment by emile0909** [comment]
If implementing teleport is relative complex, I have some thought which is inspired by how VRChat teleports and I myself feel comfortable. [VRChat teleport](https://www.youtube.com/watch?v=zMMLjbe7RQU&t=53s)

We can have a key bind to switch between normal mode and teleport mode. When user is in teleport mode, if user pushes left joystick, the camera position freezes; if user releases left joystick, camera position doesn't freeze and then teleports to the new location.

I suppose this need less effort to experiment and we can see if user is satisfied with this work around.

**Comment by carlobs** [comment]
Unfortunately I don't use UEVR because I suffer from nausea without Teleport... so I hope this feature will open UEVR to many players

**Comment by Altrue** [comment]
Upvoting this suggestion, and the VRChat idea is so cool as well!

VR accessibility is heavily limited by people who can't handle smooth locomotion, and the ability to teleport completely nullify the issue.

Alternatively, a stopgap solution that _actually works_, inspired by Until You Fall (and also Skyrim VR I believe?) is to heavily reduce the field of view while movement is occuring. I'm obviously not taking about a zoom, but rather a vignette effect that completely obscures everything except a circular area at the center of the screen.

**Comment by witan777** [comment]
Yeah, UEVR still can’t be used by people who suffer from motion sickness unless they can use instant movement (teleport).
Don’t think that upvoting will help, but leaving this comment just in case.
